### PR TITLE
add CURL login options for IMAP

### LINKED ISSUE #153: IMAP - can not chose compatible AUTH backend
### Steps to reproduce
1. Setup IMAP auth with TLS enabled to server with AUTH=GSSAPI CAPABLITY

### Expected behaviour
You can login to NC

### Actual behaviour
No login - error message in log:
[user_external][3] ERROR: Could not connect to imap server via curl: Operation timed out after 10001 milliseconds with 0 out of 0 bytes received

test from Linux shell
```
# curl --basic -v imaps://email.example.com --user 'user@example.com'
Enter host password for user 'user@madeta.cz':
*   Trying xx8.yy5.zz4.aa6:993...
* TCP_NODELAY set
* Connected to email.example.com (xx8.yy5.zz4.aa6) port 993 (#0)
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* Server certificate:
*  subject: CN=email.example.com
*  start date: Jun 17 02:22:59 2020 GMT
*  expire date: Sep 18 02:22:59 2020 GMT
*  subjectAltName: host "email.example.com" matched cert's "email.example.com"
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< * OK Server 1 IMAP4rev1 Thu, 09 Jul 2020 11:08:40 +0200
> A001 CAPABILITY
< * CAPABILITY IMAP4rev1 AUTH=PLAIN AUTH=LOGIN AUTH=DIGEST-MD5 AUTH=CRAM-MD5 AUTH=NTLM AUTH=GSSAPI SORT THREAD=ORDEREDSUBJECT UIDPLUS QUOTA ACL NAMESPACE CHILDREN IDLE ID UNSELECT METADATA X-ICEWARP-SERVER X-MOVE MULTISEARCH ESEARCH XLIST CREATE-SPECIAL-USE 
< A001 OK CAPABILITY Completed
> A002 AUTHENTICATE GSSAPI
< +
* gss_init_sec_context() failed: No Kerberos credentials available (default cache: FILE:/tmp/krb5cc_0). 
* Closing connection 0
* TLSv1.3 (OUT), TLS alert, close notify (256):
curl: (94) An authentication function returned an error
```
Problem is that this Linux server is (and will never be) used in Kerberos environment ...

**I had to switch off GSSAPI auth backend on IMAP server, because there is not any way how to tell  
user_external/CURL not to use it.**
**After it it works like charm ...***
**BTW in owncloud with the old way to specify user_external config it still works**

Possible solutions:
- enable choosing auth backend in user_external (CURLOPT_HTTPAUTH 	The HTTP authentication method(s) to use. The options are: CURLAUTH_BASIC, CURLAUTH_DIGEST, CURLAUTH_GSSNEGOTIATE, CURLAUTH_NTLM, CURLAUTH_ANY, and CURLAUTH_ANYSAFE. )
- try to auth with other auth backends too ...

### Affected Authentication backend
IMAP and server with GSSAPI

### Server configuration

**User External App version:** 0.10.0

**Operating system:** Ubuntu 20.04 LTS

**Web server:** Apache2

**Database:** Postgresql 12.2

**PHP version:** 7.4.3

**Nextcloud version:** 19.0.0.12

**Updated from an older Nextcloud/ownCloud or fresh install:** fresh 

**Where did you install Nextcloud from:** nextcloud tar file

**Signing status:**
<details>
<summary>Signing status</summary>

```
NO integrit URL ...
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
# sudo -u www-data php occ app:list
Enabled:
  - admin_audit: 1.9.0
  - cloud_federation_api: 1.2.0
  - comments: 1.9.0
  - dav: 1.15.0
  - federatedfilesharing: 1.9.0
  - files: 1.14.0
  - files_rightclick: 0.16.0
  - files_sharing: 1.11.0
  - files_trashbin: 1.9.0
  - files_versions: 1.12.0
  - files_videoplayer: 1.8.0
  - logreader: 2.4.0
  - lookup_server_connector: 1.7.0
  - notifications: 2.7.0
  - oauth2: 1.7.0
  - password_policy: 1.9.1
  - photos: 1.1.0
  - provisioning_api: 1.9.0
  - serverinfo: 1.9.0
  - settings: 1.1.0
  - sharebymail: 1.9.0
  - text: 3.0.1
  - twofactor_backupcodes: 1.8.0
  - updatenotification: 1.9.0
  - user_external: 0.10.0
  - viewer: 1.3.0
  - workflowengine: 2.1.0
Disabled:
  - accessibility
  - activity
  - contactsinteraction
  - encryption
  - federation
  - files_external
  - files_pdfviewer
  - firstrunwizard
  - limit_login_to_ip
  - nextcloud_announcements
  - privacy
  - recommendations
  - support
  - survey_client
  - systemtags
  - theming
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
# sudo -u www-data php occ config:list system
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.example.com"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "19.0.0.12",
        "overwrite.cli.url": "https:\/\/cloud.example.com",
        "htaccess.RewriteBase": "\/",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_smtpmode": "smtp",
        "mail_smtpsecure": "tls",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "user_backends": [
            {
                "class": "OC_User_IMAP",
                "arguments": [
                    "email.example.com",
                    993,
                    true
                ]
            }
        ],
        "loglevel": 3,
        "logfile": "\/var\/log\/nextcloud.log",
        "default_language": "cs_CZ",
        "updatechecker": true,
        "filelocking.enabled": true,
        "log_type": "errorlog",
        "trashbin_retention_obligation": "30, auto"
    }
}
```
</details>

### PR DESCRIPTION
**Pull request by qaxi**
Fixes #153 

Changes proposed in this pull request:
 - add one IMAP config parameter
 - allows to choose IMAP login options

Signed-off-by: Petr Klíma <qaxi@seznam.cz>

### CODE REVIEW

**Comment by qaxi** [comment]
@violoncelloCH sorry for closing #176, merge was too messy to continue. That is reason for opening it again as #196 .
I know v.3.0.0 has priority, nevertheless please take look at it, I think it is usefull when CURL authorization autodetection fails.

**Review by lazki — APPROVED** [review]
I can confirm this works in NC 25.0.01 and External user authentication  V 3.0.0.

Please merge in next version.

File: README.md [hunk 1]
```diff
@@ -89,7 +89,9 @@ is set to true, after successfull login the domain part will be striped and
 the rest used as username in Nextcloud. e.g. 'username@example.com' will be
 'username' in Nextcloud. The sixth parameter toggles whether on creation of
 the user, it is added to a group corresponding to the name of the domain part
-of the address. 
+of the address. The seventh parameter allows you specify - if is set - login
+options (for example, what authentication method you prefer).
```
Comment (violoncelloCH) [inline README.md:93]: could you add the possible values for that seventh parameter to the documentation here ?

**Review by violoncelloCH — COMMENTED** [review]
thank you for your contribution @qaxi ! I just have one small remark concerning the documentation

File: README.md (hunk 1 above)
Comment (qaxi) [reply README.md]: Is this OK?

**Review by lazki — APPROVED** [review]
I can confirm this works in NC 25.0.2 and External user authentication V 3.1.0.

Please merge in next version.

**Comment by lazki** [comment]
@qaxi  @violoncelloCH  please merge this PR. It's clean and simple. 

Thanks in advance.

**Comment by lazki** [comment]
@qaxi @violoncelloCH  will one day be merged?

**Comment by qaxi** [comment]
> @qaxi could you correct your signature in this commit? [0140407](https://github.com/nextcloud/user_external/commit/01404074a328a17342b5aa3a3fdcd216341869e1)
> 
> Here is how to correct it: https://github.com/nextcloud/user_external/pull/196/checks?check_run_id=10391753498

Sorry I badly failed and I am unable to do it ...

$ git push --force-with-lease origin master
X11 forwarding request failed on channel 0
ERROR: Permission to nextcloud/user_external.git denied to qaxi.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

**Pull request closed without merging**